src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1188,30 +1192,88 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
-
-#if HAVE_FCOPYFILE
-    // If fcopyfile is available (OS X), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
-    // Copy data and metadata.
-    return fcopyfile(inFd, outFd, NULL, COPYFILE_ALL) == 0 ? 0 : -1;
-#else
-    // Get the stats on the source file.
+    int outFd;
     int ret;
     struct stat_ sourceStat;
-    bool copied = false;
-#if HAVE_SENDFILE_4
-    // If sendfile is available (Linux), try to use it, as the whole copy
-    // can be performed in the kernel, without lots of unnecessary copying.
+
     while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
     if (ret != 0)
     {
         return -1;
     }
 
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
```

Comment:
Why EBUSY rather than EEXIST?

Reply:
Let me check. It's been a while but I think I was emulating the error code the previous code would return, ie. trying to open file second time with different permissions.

Reply:
It should depend on the `overwrite` parameter.

`overwrite`: `System.IO.IOException: The process cannot access the file '/tmp/tmpy5n4qV.tmp' because it is being used by another process.`
`!overwrite`: `System.IO.IOException: The file '/tmp/tmppUACtg.tmp' already exists.`